pkg/cmd/version/version.go

```diff
@@ -30,39 +30,40 @@ type VersionOptions struct {
 }
 
 func NewCmdVersion(commonOpts *opts.CommonOptions) *cobra.Command {
-	options := &VersionOptions{
+	vOpts := &VersionOptions{
 		CommonOptions: commonOpts,
 	}
 
 	cmd := &cobra.Command{
 		Use:   "version",
 		Short: "Print the version information",
 		Run: func(cmd *cobra.Command, args []string) {
-			options.Cmd = cmd
-			options.Args = args
-			err := options.Run()
+			vOpts.Cmd = cmd
+			vOpts.Args = args
+			err := vOpts.Run()
 			helper.CheckErr(err)
 		},
 	}
 	/*
 		cmd.Flags().BoolP("client", "c", false, "Client version only (no server required).")
 		cmd.Flags().BoolP("short", "", false, "Print just the version number.")
 	*/
-	cmd.Flags().MarkShorthandDeprecated("client", "please use --client instead.")
-	cmd.Flags().BoolVarP(&options.HelmTLS, "helm-tls", "", false, "Whether to use TLS with helm")
-	cmd.Flags().BoolVarP(&options.NoVersionCheck, "no-version-check", "n", false, "Disable checking of version upgrade checks")
-	cmd.Flags().BoolVarP(&options.NoVerify, "no-verify", "", false, "Disable verification of package versions")
-	cmd.Flags().StringVarP(&options.Namespace, "namespace", "", "", "The namespace to use to look for currently installed platform version")
+	_ = cmd.Flags().MarkShorthandDeprecated("client", "please use --client instead.")
+	cmd.Flags().BoolVarP(&vOpts.HelmTLS, "helm-tls", "", false, "Whether to use TLS with helm")
+	cmd.Flags().BoolVarP(&vOpts.NoVersionCheck, "no-version-check", "n", false, "Disable checking of version upgrade checks")
+	cmd.Flags().BoolVarP(&vOpts.NoVerify, "no-verify", "", false, "Disable verification of package versions")
+	cmd.Flags().StringVarP(&vOpts.Namespace, "namespace", "", "", "The namespace to use to look for currently installed platform version")
 	return cmd
 }
 
 func (o *VersionOptions) Run() error {
 	ns := o.Namespace
+	fmt.Println("Namespace: ", ns)
 	if ns == "" {
 		var err error
 		_, ns, err = o.JXClientAndDevNamespace()
 		if err != nil {
-			return err
+			log.Logger().Warnf("Error with jx client: %v", err)
```

Comment:
I am not sure whether this is the right thing here to do since the namespace is used in `GetPackageVersions`. The latter really should be called against the dev namespace of a Jenkins X install. With your change we pass the empty namespace. This is at the very least changing intent. 

For all I can tell, it also does not matter if I call `jx version` from a non jx namespace, as long as there is a Jenkins X system namespace at all. Problematic is to be connected to an invalid context or a cluster w/o Jenkins X installation.

There is an ambiguity here which I would love to resolve, but I think it would need more doing and a discussion across the team on what to do. There are two commands `jx --version` and `jx version`. The former literally just returns the version of the `jx` binary. No connection to the cluster. Simple. `jx version` on the other hand does more. It does report the version of the binary, but then it also checks for the cluster, git version, etc. A lot of users get caught by that.

I think to really "fix" that we need to change the behaviour of `jx version` to be the same as `jx --version`. We can then either deprecate/remove `jx --version` in the long run or just keep both. I think in the space we are with `jx` (helm, kustomize, cloud, ...) most people expect the version command to just return the version of the binary. 

The current functionality of `jx version` is basically a duplicate or at least very similar to `jx diagnose` (which might need some updating). That's what I would refer people to use instead of the current `jx version`.

@jenkins-x/core thoughts?

Reply:
I think it should mimic what `kubectl version` and `helm version` do. They just print the client and server versions and exit. Version should not do more than that imo. The issue I had was it failed to print anything when there was an invalid namespace, which is a bit odd, as `kubectl version` does not throw error for that use case. 
I agree that version does a lot more than what it should do, just this much information should get printed imo.
```
NAME               VERSION
jx                 2.0.1065-dev+52231531f
Kubernetes cluster v1.15.6
kubectl            v1.16.2
```

Reply:
I am not so sure about `Kubernetes cluster` and `kubectel`. With the former, you won't know whether the cluster which it reports is the actual Jenkins X cluster. Your current context might be set to something else. Also, it does imo not provide much useful context for `jx`. Same for `kubectl`, not sure why I need the version for that. But trying to do less is already a step into the right direction.

Reply:
Right, I agree. We should move:
```
Kubernetes cluster v1.15.6
kubectl            v1.16.2
```
to `jx diagnose`. And make it like `helm version` (v2). Print the binary information, and the server side information. That would be ideal imo

Reply:
> Right, I agree. We should move:

> Kubernetes cluster v1.15.6
> kubectl            v1.16.2
> to jx diagnose

+1, that makes more sense imo. Diagnose is where you try to collect as much useful information as possible.

> Print the binary information, and the server-side information. That would be ideal imo.

Right, but I am not sure what we have atm in terms of determining the server-side version of Jenkins X. And of course this it would require to find the Jenkins X namespace. I guess you would print the client-side version and then try to get some information about the server-side and if for some reasons, not possible print a warning of sorts. Similar to what Helm 2 does with Tiller. Persoanlly, I could live with just the client-side version for now.

> As an example, say the current context is set incorrectly, then the behavior should be:

> client: <jx-binary-version>
> server: <err message/warning message>

Exactly.

Reply:
I am also thinking that client side version should be sufficient, there is no associated version for jx namespace anyways. We should remove the `--version` option to make it consistent with the other kubernetes cli tools imo. wdyt?

Reply:
> We should remove the --version option to make it consistent with the other kubernetes cli tools imo. wdyt

I guess we would deprecate it first, opposed to removing it directly.

Reply:
So a plan of action could be:

1. Fix `jx diagnose` and ensure it contains all output we want. I would think it might make sense to have something like `jx diagnose --long` or similar which when enabled runs the additional `kubectl` commands which are currently in `jx diagnose`. Also, looking at the current implementation of `jx diagnose`, it uses os.exec to call `jx` commands itself. That seems wrong and depending on your PATH can have unwanted results. Might be worth do tackle `jx diagnose` in a dedicated issue which outlines the planned changes.
1. Once `jx diagnose` behaves as described above, we can change `jx version` to just return the version of the CLI. This is inline with other tools in the K8S space
1. Deprecate `jx --version` and eventually remove it

Reply:
Makes sense to deprecate it first, and then remove it 👍

Reply:
I will open an issue for jx diagnose, and a PR for that. For the version command, this is what I am thinking:
- `jx version --short`:
```
version: v1.16.2
``` 
- `jx version`:
```
version: X.Y.Z
commit: commitid
git tree state: always clean!
```
Pretty much what the other K8s tools do :), wdyt?

Reply:
I am most interested in the short version, but nothing if it is not too complicated to make your suggestion work I am fine with it as well. For development versions, we already add the commit sha to the version - https://github.com/jenkins-x/jx/blob/22ce2138110e61c4497f02efc8205dd2fc8de718/Makefile#L33. Looks like this `2.1.33-dev+343f641ad`. To do what you are suggesting we need to change this a bit and always (even for release builds where there is an explicit VERSION set) compile in the build SHA.  

Regarding `git tree state`, I am not even really sure what that means. I guess that's whether you compile the binary with outstanding changes in the working tree. Not sure how valuable that is in the end. 

Important for me is that the version command is not doing network calls and is independent of your current Kube context. All information should be really be compiled into the binary.

Reply:
@ankitm123 given our latest plans what shall we do with this pull request? Keep it open or close for now?

Reply:
I think we should keep it open, I started refactoring `jx diagnose` in PR #7178. Once that is completed and merged, I will complete this. I did remove the code which was doing all the extra stuff in this PR (communicating to the kubernetes cluster, printing kubectl versions etc), but as we decided, we refactor jx diagnose, and then once that PR is merged, we merge this PR, wdyt?

Reply:
Might be worth playing a bit with the `diagnose` options. Not sure whether we always want to print all the pvcs, pods, etc.

Reply:
Thought abt the same thing, there is no way to only print version or configmaps or status using `jx diagnose`, making those `o.showOption("pods")` redundant.  I was thinking something along the lines of `jx diagnose <resource>`. I will be updating #7192 this weekend.

Reply:
@hferentschik the only thing left in this PR is to figure out how to mark `--version` as deprecated. I found the place where we set it in the codebase (cmd.go), 
```
rootCommand.Version = version.GetVersion()
rootCommand.SetVersionTemplate("{{printf .Version}}\n")
```
but seems like 
```
cmd.Flags().MarkDeprecated("version", "")
```
does not work for this flag. I am trying to figure that out. As a hack, I can always add that deprecation information to the version template, but I would not prefer not to (it's ugly).